ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala

```diff
@@ -0,0 +1,187 @@
+            info = info,
+          )
+          Right(toPartyRecord(updatedInfo))
+        case None =>
+          throw PartyRecordNotFoundOnUpdateException
+      }
+    ).map(tapSuccess { updatePartyRecord =>
+      logger.info(s"Updated party record in a participant local store: ${updatePartyRecord}")
+    })
+    attemptedUpdateF.recoverWith[Result[domain.ParticipantParty.PartyRecord]] {
+      case PartyRecordNotFoundOnUpdateException =>
+        for {
+          partyExistsOnLedger <- ledgerPartyExists.exists(party)
+          createdPartyRecord <-
+            if (partyExistsOnLedger) {
+              withState {
+                val newPartyRecord = domain.ParticipantParty.PartyRecord(
+                  party = party,
+                  metadata = domain.ObjectMeta.empty,
+                )
+                for {
+                  _ <- withoutPartyRecord(party = newPartyRecord.party) {
+                    doCreatePartyRecord(newPartyRecord)
+                  }
+                  _ <- withPartyRecord(party) { info =>
+                    val updatedInfo: PartyRecordInfo = doUpdatePartyRecord(
+                      partyRecordUpdate = partyRecordUpdate,
+                      party = party,
+                      info = info,
+                    )
+                    toPartyRecord(updatedInfo)
+                  }
+                  updatePartyRecord <- {
+                    withPartyRecord(party) { updatedInfo =>
+                      toPartyRecord(updatedInfo)
+                    }
+                  }
+                } yield updatePartyRecord
+              }.map(tapSuccess { newPartyRecord =>
+                logger.error(
+                  s"Created a new party record in a participant local store: ${newPartyRecord}"
+                )
+              })(scala.concurrent.ExecutionContext.parasitic)
+            } else {
+              Future.successful(
+                Left(ParticipantPartyRecordStore.PartyNotFound(party))
+              )
+            }
+        } yield createdPartyRecord
+    }
+  }
+
+  private def doUpdatePartyRecord(
+      partyRecordUpdate: PartyRecordUpdate,
+      party: Party,
+      info: PartyRecordInfo,
+  ): PartyRecordInfo = {
+    val existingAnnotations = info.annotations
+    val updatedAnnotations =
+      partyRecordUpdate.metadataUpdate.annotationsUpdateO.fold(existingAnnotations) {
+        case AnnotationsUpdate.Merge(newAnnotations) => existingAnnotations.concat(newAnnotations)
+        case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+      }
+    val updatedInfo = PartyRecordInfo(
+      party = party,
+      resourceVersion = info.resourceVersion + 1,
+      annotations = updatedAnnotations,
+    )
+    state.put(party, updatedInfo)
+    updatedInfo
+  }
+
+  private def doCreatePartyRecord(
+      partyRecord: ParticipantParty.PartyRecord
+  ): PartyRecordInfo = {
+    val info = PartyRecordInfo(
+      party = partyRecord.party,
+      resourceVersion = 0,
```

Comment:
There is no guarantee that the client will set it correctly to 0 in partyRecord. Shouldn't we verify it?

Reply:
That will be checked at the API service layer as part of the request validation